File: init_firestore.js
/**
 * Initialize and Seed Cloud Firestore for Lewi House using pure Node.js built-ins.
 * (No npm install required - uses built-in crypto and https modules).
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const crypto = require('crypto');

// 1. Locate the Service Account JSON key
const possibleKeyFiles = [
  'lewihouse-7a0d7-firebase-adminsdk-fbsvc-95805aa62a.json',
  'serviceAccountKey.json',
  ...fs.readdirSync(__dirname).filter(f => f.includes('firebase-adminsdk') && f.endsWith('.json'))
];

let keyFilePath = possibleKeyFiles.find(f => fs.existsSync(path.join(__dirname, f)));
if (!keyFilePath) {
  console.error('[!] No Firebase service account JSON key found in workspace.');
  process.exit(1);
}

const keyPath = path.join(__dirname, keyFilePath);
const sa = JSON.parse(fs.readFileSync(keyPath, 'utf8'));
const PROJECT_ID = sa.project_id || 'lewihouse-7a0d7';

console.log(`[*] Target Firebase Project ID: ${PROJECT_ID}`);
console.log(`[*] Using Service Account Key: ${keyFilePath}`);

// 2. Generate Google OAuth2 Access Token using RS256 JWT
function getAccessToken() {
  return new Promise((resolve, reject) => {
    const now = Math.floor(Date.now() / 1000);
    const header = { alg: 'RS256', typ: 'JWT' };
    const claimSet = {
      iss: sa.client_email,
      scope: 'https://www.googleapis.com/auth/datastore https://www.googleapis.com/auth/cloud-platform',
      aud: 'https://oauth2.googleapis.com/token',
      exp: now + 3600,
      iat: now,
    };

    const b64Url = (obj) => Buffer.from(JSON.stringify(obj)).toString('base64url');
    const unsignedJwt = `${b64Url(header)}.${b64Url(claimSet)}`;

    const sign = crypto.createSign('RSA-SHA256');
    sign.update(unsignedJwt);
    sign.end();
    const signature = sign.sign(sa.private_key, 'base64url');
    const assertion = `${unsignedJwt}.${signature}`;

    const postData = new URLSearchParams({
      grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
      assertion: assertion,
    }).toString();

    const req = https.request('https://oauth2.googleapis.com/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(postData),
      },
    }, (res) => {
      let data = '';
      res.on('data', (c) => data += c);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          if (json.access_token) {
            resolve(json.access_token);
          } else {
            reject(new Error(JSON.stringify(json)));
          }
        } catch (e) {
          reject(e);
        }
      });
    });

    req.on('error', reject);
    req.write(postData);
    req.end();
  });
}

// 3. Firestore REST Value Serializer
function toFirestoreValue(val) {
  if (val === null || val === undefined) return { nullValue: null };
  if (typeof val === 'boolean') return { booleanValue: val };
  if (typeof val === 'number') {
    return Number.isInteger(val) ? { integerValue: String(val) } : { doubleValue: val };
  }
  if (typeof val === 'string') return { stringValue: val };
  if (Array.isArray(val)) {
    return { arrayValue: { values: val.map(toFirestoreValue) } };
  }
  if (typeof val === 'object') {
    const fields = {};
    for (const [k, v] of Object.entries(val)) {
      if (k !== '_id') fields[k] = toFirestoreValue(v);
    }
    return { mapValue: { fields } };
  }
  return { stringValue: String(val) };
}

function writeFirestoreDoc(token, docPath, docData) {
  return new Promise((resolve, reject) => {
    const fields = {};
    for (const [k, v] of Object.entries(docData)) {
      if (k !== '_id') fields[k] = toFirestoreValue(v);
    }
    const body = JSON.stringify({ fields });
    const url = `/v1/projects/${PROJECT_ID}/databases/(default)/documents/${docPath}`;

    const req = https.request({
      hostname: 'firestore.googleapis.com',
      path: url,
      method: 'PATCH',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
      }
    }, (res) => {
      let data = '';
      res.on('data', (c) => data += c);
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(true);
        } else {
          console.warn(`  [-] HTTP ${res.statusCode} for ${docPath}: ${data}`);
          resolve(false);
        }
      });
    });

    req.on('error', (err) => {
      console.error(`  [!] Request error for ${docPath}: ${err.message}`);
      resolve(false);
    });

    req.write(body);
    req.end();
  });
}

async function main() {
  try {
    const token = await getAccessToken();
    console.log('  [✓] Successfully generated Google OAuth2 Token from Service Account.');

    const now = new Date().toISOString();
    const PROPERTY_SCOPE = 'properties/lewi_house_main';

    console.log('\n[*] Initializing Firestore collections...');

    // 1. Property Info
    const propertyInfo = {
      id: 'lewi_house_main',
      name: 'Lewi House Medan',
      legalName: 'Lewi House Syariah / Kost Lewi House',
      address: 'Jl. Sei Bahkapuran No. 16A, Sei Sikambing D, Kec. Medan Petisah, Kota Medan, Sumatera Utara 20119',
      totalFloors: 4,
      totalRooms: 17,
      buildingType: 'Kost Campur Eksklusif & Guesthouse Syariah',
      contactPhone: '+62 812-6296-0211',
      contactEmail: 'fauziealiakhmad@gmail.com',
      wifiSsid: 'LewiHouse_Guest',
      onSiteServices: ['LEWI Laundry', 'Area Parkir Mobil & Motor', 'Rooftop Terrace & Workout Area'],
      checkInTime: '14:00',
      checkOutTime: '12:00',
      updatedAt: now,
    };
    await writeFirestoreDoc(token, 'properties/lewi_house_main', propertyInfo);
    console.log('  [✓] Property document "lewi_house_main" initialized (Lewi House Medan).');

    // 2. Kosan Rooms (17 Kamar across 4 Lantai)
    const rooms = [
      // Lantai 1 (4 Kamar)
      { id: 'room_101', roomNumber: '101', floor: '1', roomType: 'tipe_a', capacity: 2, monthlyPrice: 2500000, deposit: 1000000, status: 'OCCUPIED', photoUrl: '/gallery/agoda/agoda-10-deluxe-bed.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Queen', 'Meja Kerja', 'Lemari Cermin', 'Smart TV', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_102', roomNumber: '102', floor: '1', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2000000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-15-superior-double.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Single', 'Meja', 'Lemari', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_103', roomNumber: '103', floor: '1', roomType: 'tipe_c', capacity: 1, monthlyPrice: 1600000, deposit: 500000, status: 'OCCUPIED', photoUrl: '/gallery/agoda/agoda-11-standard-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur 1.2x2m', 'Meja Kompak', 'Lemari', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_104', roomNumber: '104', floor: '1', roomType: 'tipe_c', capacity: 1, monthlyPrice: 1600000, deposit: 500000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-13-standard-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur 1.2x2m', 'Meja Kompak', 'Lemari', 'WiFi Cepat'], updatedAt: now },

      // Lantai 2 (5 Kamar)
      { id: 'room_201', roomNumber: '201', floor: '2', roomType: 'tipe_a', capacity: 2, monthlyPrice: 2500000, deposit: 1000000, status: 'OCCUPIED', photoUrl: '/gallery/agoda/agoda-02-suite-bedroom.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Queen', 'Meja Kerja', 'Lemari Cermin', 'Smart TV', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_202', roomNumber: '202', floor: '2', roomType: 'tipe_a', capacity: 2, monthlyPrice: 2500000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-09-suite-bedroom.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Queen', 'Meja Kerja', 'Lemari Cermin', 'Smart TV', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_203', roomNumber: '203', floor: '2', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2000000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-12-superior-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Single', 'Meja', 'Lemari', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_204', roomNumber: '204', floor: '2', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2000000, deposit: 1000000, status: 'OCCUPIED', photoUrl: '/gallery/agoda/agoda-15-superior-double.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Double', 'Meja', 'Lemari', 'TV', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_205', roomNumber: '205', floor: '2', roomType: 'tipe_c', capacity: 1, monthlyPrice: 1600000, deposit: 500000, status: 'CLEANING', photoUrl: '/gallery/agoda/agoda-11-standard-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur 1.2x2m', 'Meja Kompak', 'Lemari', 'WiFi Cepat'], updatedAt: now },

      // Lantai 3 (4 Kamar)
      { id: 'room_301', roomNumber: '301', floor: '3', roomType: 'tipe_a', capacity: 2, monthlyPrice: 2500000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-10-deluxe-bed.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Queen', 'Meja Kerja', 'Lemari Cermin', 'Smart TV', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_302', roomNumber: '302', floor: '3', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2000000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-14-deluxe-bedroom.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Single', 'Meja', 'Lemari', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_303', roomNumber: '303', floor: '3', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2000000, deposit: 1000000, status: 'OCCUPIED', photoUrl: '/gallery/agoda/agoda-15-superior-double.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Double', 'Meja', 'Lemari', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_304', roomNumber: '304', floor: '3', roomType: 'tipe_c', capacity: 1, monthlyPrice: 1600000, deposit: 500000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-13-standard-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur 1.2x2m', 'Meja Kompak', 'Lemari', 'WiFi Cepat'], updatedAt: now },

      // Lantai 4 (4 Kamar)
      { id: 'room_401', roomNumber: '401', floor: '4', roomType: 'tipe_a', capacity: 2, monthlyPrice: 2600000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-02-suite-bedroom.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Queen', 'Meja Kerja', 'Lemari Cermin', 'Smart TV', 'WiFi Cepat', 'Akses Rooftop'], updatedAt: now },
      { id: 'room_402', roomNumber: '402', floor: '4', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2100000, deposit: 1000000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-12-superior-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Single', 'Meja', 'Lemari', 'WiFi Cepat', 'Akses Rooftop'], updatedAt: now },
      { id: 'room_403', roomNumber: '403', floor: '4', roomType: 'tipe_b', capacity: 1, monthlyPrice: 2100000, deposit: 1000000, status: 'MAINTENANCE', photoUrl: '/gallery/agoda/agoda-14-deluxe-bedroom.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur Springbed Single', 'Meja', 'Lemari', 'WiFi Cepat'], updatedAt: now },
      { id: 'room_404', roomNumber: '404', floor: '4', roomType: 'tipe_c', capacity: 1, monthlyPrice: 1700000, deposit: 500000, status: 'AVAILABLE', photoUrl: '/gallery/agoda/agoda-11-standard-single.webp', facilities: ['AC', 'Kamar Mandi Dalam (Shower & Kloset Duduk)', 'Kasur 1.2x2m', 'Meja Kompak', 'Lemari', 'WiFi Cepat'], updatedAt: now },
    ];

    for (const r of rooms) {
      await writeFirestoreDoc(token, `${PROPERTY_SCOPE}/rooms/${r.id}`, r);
    }
    console.log(`  [✓] ${rooms.length} Rooms initialized (101-104, 201-204).`);

    // 3. Demo Resident
    const resident = {
      id: 'resident_101',
      fullName: 'Budi Santoso',
      email: 'budi.santoso@example.com',
      phone: '+6281298765432',
      roomNumber: '101',
      moveInDate: '2026-01-01',
      leaseEndDate: '2026-12-31',
      monthlyRent: 1500000,
      depositAmount: 500000,
      status: 'ACTIVE',
      emergencyContact: 'Orang Tua',
      emergencyPhone: '+628111222333',
      ktpNumber: '3201234567890001',
      updatedAt: now,
    };
    await writeFirestoreDoc(token, `${PROPERTY_SCOPE}/residents/${resident.id}`, resident);
    console.log('  [✓] Resident "Budi Santoso" seeded.');

    // 4. Admin User
    const adminUser = {
      id: 'usr_owner_1',
      email: 'fauziealiakhmad@gmail.com',
      name: 'Fauzie Ali Akhmad',
      role: 'owner',
      createdAt: now,
    };
    await writeFirestoreDoc(token, 'users/usr_owner_1', adminUser);
    console.log('  [✓] Owner/Admin profile (fauziealiakhmad@gmail.com) created.');

    // 5. Electricity Meters
    for (const r of rooms) {
      const meter = {
        id: `meter_${r.roomNumber}`,
        roomNumber: r.roomNumber,
        meterNumber: `54129800${r.roomNumber}`,
        currentKwh: 45.8,
        lastUpdated: now,
      };
      await writeFirestoreDoc(token, `${PROPERTY_SCOPE}/electricity_meters/${meter.id}`, meter);
    }
    console.log('  [✓] 8 Electricity meters initialized.');

    // 6. Announcements
    const notif = {
      id: 'notif_welcome',
      title: 'Selamat Datang di Lewi House',
      body: 'Aplikasi manajemen kosan Lewi House kini aktif. Seluruh data tagihan dan hunian siap digunakan.',
      category: 'ANNOUNCEMENT',
      recipientResidentId: null,
      createdAt: now,
      isRead: false,
    };
    await writeFirestoreDoc(token, `${PROPERTY_SCOPE}/notifications/${notif.id}`, notif);
    console.log('  [✓] Welcome notification announcement created.');

    console.log('\n' + '='.repeat(50));
    console.log('🎉 CLOUD FIRESTORE SETUP & SEEDING COMPLETE!');
    console.log('='.repeat(50));
  } catch (err) {
    console.error('[!] Execution failed:', err);
  }
}

main();

File: init_firestore.py
"""Initialize and Seed Cloud Firestore for Lewi House using Service Account Credentials."""

import os
import sys
import json
import urllib.request
import urllib.error
from pathlib import Path
from datetime import datetime, timezone

# Locate serviceAccountKey.json or downloaded firebase-adminsdk json
KEY_PATHS = [
    Path("serviceAccountKey.json"),
    Path("backend/serviceAccountKey.json"),
    *list(Path(".").glob("*firebase-adminsdk*.json")),
    *list(Path("backend").glob("*firebase-adminsdk*.json")),
    *list((Path.home() / "Downloads").glob("*firebase-adminsdk*.json")),
]

service_key_path = None
for p in KEY_PATHS:
    if p.exists():
        service_key_path = p
        break

if not service_key_path:
    print("[!] No service account json found.")
    sys.exit(1)

with open(service_key_path, "r", encoding="utf-8") as f:
    sa_info = json.load(f)

PROJECT_ID = sa_info.get("project_id", "lewihouse-7a0d7")
print(f"[*] Target Firebase Project ID: {PROJECT_ID}")
print(f"[*] Using Service Account Key: {service_key_path.name}")

# Obtain OAuth2 Bearer Token using google-auth or pyjwt/cryptography
try:
    from google.oauth2 import service_account
    import google.auth.transport.requests

    SCOPES = ["https://www.googleapis.com/auth/datastore"]
    creds = service_account.Credentials.from_service_account_file(
        str(service_key_path), scopes=SCOPES
    )
    request = google.auth.transport.requests.Request()
    creds.refresh(request)
    ACCESS_TOKEN = creds.token
    print("  [✓] Successfully authenticated Service Account and generated access token.")
except Exception as e:
    print(f"[!] Failed to get token with google-auth: {e}")
    sys.exit(1)

FIRESTORE_BASE = f"https://firestore.googleapis.com/v1/projects/{PROJECT_ID}/databases/(default)/documents"
PROPERTY_SCOPE = "properties/lewi_house_main"
now = datetime.now(timezone.utc).isoformat()


def to_firestore_value(val):
    if val is None:
        return {"nullValue": None}
    elif isinstance(val, bool):
        return {"booleanValue": val}
    elif isinstance(val, int):
        return {"integerValue": str(val)}
    elif isinstance(val, float):
        return {"doubleValue": float(val)}
    elif isinstance(val, str):
        return {"stringValue": val}
    elif isinstance(val, list):
        return {"arrayValue": {"values": [to_firestore_value(v) for v in val]}}
    elif isinstance(val, dict):
        return {"mapValue": {"fields": {k: to_firestore_value(v) for k, v in val.items()}}}
    return {"stringValue": str(val)}


def to_firestore_fields(data):
    fields = {}
    for k, v in data.items():
        if k == "_id":
            continue
        fields[k] = to_firestore_value(v)
    return fields


def write_doc(path: str, data: dict) -> bool:
    url = f"{FIRESTORE_BASE}/{path}"
    payload = {"fields": to_firestore_fields(data)}
    req = urllib.request.Request(url, method="PATCH")
    req.add_header("Authorization", f"Bearer {ACCESS_TOKEN}")
    req.add_header("Content-Type", "application/json")
    data_bytes = json.dumps(payload).encode("utf-8")
    try:
        with urllib.request.urlopen(req, data=data_bytes, timeout=15) as resp:
            return resp.status in (200, 201)
    except urllib.error.HTTPError as err:
        err_msg = err.read().decode("utf-8", errors="ignore")
        print(f"  [-] Failed writing {path}: HTTP {err.code} - {err_msg}")
        return False
    except Exception as ex:
        print(f"  [-] Error on {path}: {ex}")
        return False


def seed_all():
    print("\n[*] Populating Firestore database collections...")

    # 1. Main Property Doc
    property_info = {
        "id": "lewi_house_main",
        "name": "Lewi House Medan",
        "legalName": "Lewi House Syariah / Kost Lewi House",
        "address": "Jl. Sei Bahkapuran No. 16A, Sei Sikambing D, Kec. Medan Petisah, Kota Medan, Sumatera Utara 20119",
        "totalFloors": 4,
        "totalRooms": 17,
        "buildingType": "Kost Campur Eksklusif & Guesthouse Syariah",
        "contactPhone": "+62 812-6296-0211",
        "contactEmail": "fauziealiakhmad@gmail.com",
        "wifiSsid": "LewiHouse_Guest",
        "onSiteServices": ["LEWI Laundry", "Area Parkir Mobil & Motor", "Rooftop Terrace & Workout Area"],
        "checkInTime": "14:00",
        "checkOutTime": "12:00",
        "updatedAt": now,
    }
    if write_doc("properties/lewi_house_main", property_info):
        print("  [✓] Property document 'lewi_house_main' created (Lewi House Medan).")

    # 2. Rooms (17 Kamar across 4 Lantai)
    rooms = [
        # Lantai 1 (4 Kamar)
        {"id": "room_101", "roomNumber": "101", "floor": "1", "roomType": "tipe_a", "capacity": 2, "monthlyPrice": 2500000.0, "deposit": 1000000.0, "status": "OCCUPIED", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Queen", "Meja Kerja", "Lemari Cermin", "Smart TV", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_102", "roomNumber": "102", "floor": "1", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2000000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Single", "Meja", "Lemari", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_103", "roomNumber": "103", "floor": "1", "roomType": "tipe_c", "capacity": 1, "monthlyPrice": 1600000.0, "deposit": 500000.0, "status": "OCCUPIED", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur 1.2x2m", "Meja Kompak", "Lemari", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_104", "roomNumber": "104", "floor": "1", "roomType": "tipe_c", "capacity": 1, "monthlyPrice": 1600000.0, "deposit": 500000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur 1.2x2m", "Meja Kompak", "Lemari", "WiFi Cepat"], "updatedAt": now},

        # Lantai 2 (5 Kamar)
        {"id": "room_201", "roomNumber": "201", "floor": "2", "roomType": "tipe_a", "capacity": 2, "monthlyPrice": 2500000.0, "deposit": 1000000.0, "status": "OCCUPIED", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Queen", "Meja Kerja", "Lemari Cermin", "Smart TV", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_202", "roomNumber": "202", "floor": "2", "roomType": "tipe_a", "capacity": 2, "monthlyPrice": 2500000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Queen", "Meja Kerja", "Lemari Cermin", "Smart TV", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_203", "roomNumber": "203", "floor": "2", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2000000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Single", "Meja", "Lemari", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_204", "roomNumber": "204", "floor": "2", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2000000.0, "deposit": 1000000.0, "status": "OCCUPIED", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Double", "Meja", "Lemari", "TV", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_205", "roomNumber": "205", "floor": "2", "roomType": "tipe_c", "capacity": 1, "monthlyPrice": 1600000.0, "deposit": 500000.0, "status": "CLEANING", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur 1.2x2m", "Meja Kompak", "Lemari", "WiFi Cepat"], "updatedAt": now},

        # Lantai 3 (4 Kamar)
        {"id": "room_301", "roomNumber": "301", "floor": "3", "roomType": "tipe_a", "capacity": 2, "monthlyPrice": 2500000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Queen", "Meja Kerja", "Lemari Cermin", "Smart TV", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_302", "roomNumber": "302", "floor": "3", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2000000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Single", "Meja", "Lemari", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_303", "roomNumber": "303", "floor": "3", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2000000.0, "deposit": 1000000.0, "status": "OCCUPIED", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Double", "Meja", "Lemari", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_304", "roomNumber": "304", "floor": "3", "roomType": "tipe_c", "capacity": 1, "monthlyPrice": 1600000.0, "deposit": 500000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur 1.2x2m", "Meja Kompak", "Lemari", "WiFi Cepat"], "updatedAt": now},

        # Lantai 4 (4 Kamar)
        {"id": "room_401", "roomNumber": "401", "floor": "4", "roomType": "tipe_a", "capacity": 2, "monthlyPrice": 2600000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Queen", "Meja Kerja", "Lemari Cermin", "Smart TV", "WiFi Cepat", "Akses Rooftop"], "updatedAt": now},
        {"id": "room_402", "roomNumber": "402", "floor": "4", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2100000.0, "deposit": 1000000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Single", "Meja", "Lemari", "WiFi Cepat", "Akses Rooftop"], "updatedAt": now},
        {"id": "room_403", "roomNumber": "403", "floor": "4", "roomType": "tipe_b", "capacity": 1, "monthlyPrice": 2100000.0, "deposit": 1000000.0, "status": "MAINTENANCE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur Springbed Single", "Meja", "Lemari", "WiFi Cepat"], "updatedAt": now},
        {"id": "room_404", "roomNumber": "404", "floor": "4", "roomType": "tipe_c", "capacity": 1, "monthlyPrice": 1700000.0, "deposit": 500000.0, "status": "AVAILABLE", "facilities": ["AC", "Kamar Mandi Dalam (Shower & Kloset Duduk)", "Kasur 1.2x2m", "Meja Kompak", "Lemari", "WiFi Cepat"], "updatedAt": now},
    ]
    for r in rooms:
        write_doc(f"{PROPERTY_SCOPE}/rooms/{r['id']}", r)
    print(f"  [✓] {len(rooms)} Rooms seeded into properties/lewi_house_main/rooms.")

    # 3. Demo Resident
    resident = {
        "id": "resident_101",
        "fullName": "Budi Santoso",
        "email": "budi.santoso@example.com",
        "phone": "+6281298765432",
        "roomNumber": "101",
        "moveInDate": "2026-01-01",
        "leaseEndDate": "2026-12-31",
        "monthlyRent": 1500000.0,
        "depositAmount": 500000.0,
        "status": "ACTIVE",
        "emergencyContact": "Orang Tua",
        "emergencyPhone": "+628111222333",
        "ktpNumber": "3201234567890001",
        "updatedAt": now,
    }
    write_doc(f"{PROPERTY_SCOPE}/residents/{resident['id']}", resident)
    print("  [✓] Demo resident seeded into properties/lewi_house_main/residents.")

    # 4. Admin user profile in root users
    admin_user = {
        "id": "usr_owner_1",
        "email": "fauziealiakhmad@gmail.com",
        "name": "Fauzie Ali Akhmad",
        "role": "owner",
        "createdAt": now,
    }
    write_doc("users/usr_owner_1", admin_user)
    print("  [✓] Admin user profile registered in users collection.")

    # 5. Electricity meters
    for r in rooms:
        meter = {
            "id": f"meter_{r['roomNumber']}",
            "roomNumber": r["roomNumber"],
            "meterNumber": f"54129800{r['roomNumber']}",
            "currentKwh": 45.8,
            "lastUpdated": now,
        }
        write_doc(f"{PROPERTY_SCOPE}/electricity_meters/{meter['id']}", meter)
    print(f"  [✓] Electricity meters seeded.")

    # 6. Welcome announcement
    notif = {
        "id": "notif_welcome",
        "title": "Selamat Datang di Lewi House",
        "body": "Aplikasi manajemen kosan Lewi House kini aktif. Data dan tagihan tersinkronisasi.",
        "category": "ANNOUNCEMENT",
        "recipientResidentId": None,
        "createdAt": now,
        "isRead": False,
    }
    write_doc(f"{PROPERTY_SCOPE}/notifications/{notif['id']}", notif)
    print("  [✓] Welcome notification seeded.")

    print("\n" + "="*45)
    print("🎉 FIRESTORE DATABASE INITIALIZED & SEEDED!")
    print("="*45)


if __name__ == "__main__":
    seed_all()
